File: 필수문항/134_에너지모으기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static int N;
    public static int[] W;
    public static boolean[] visited;
    public static int ans;
    public static int num;

    public static void dfs(int energy) {
        if (num == 2) {
            ans = Math.max(ans, energy);
            return;
        }
        for (int i = 1; i < N - 1; i++) {
            if (!visited[i]) {
                num--;
                visited[i] = true;

                int lp = i - 1;
                int rp = i + 1;
                while (visited[lp]) {
                    lp--;
                }
                while (visited[rp]) {
                    rp++;
                }
                dfs(energy + W[lp] * W[rp]);

                visited[i] = false;
                num++;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        W = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        visited = new boolean[N];
        num = N;
        dfs(0);
        System.out.println(ans);
    }
}

File: 필수문항/134_에너지모으기.py
import sys

input = sys.stdin.readline
N = int(input())
W = list(map(int, input().split()))

# (1~N-2) 인덱스를 고를수있음
# N이 3일때까지만 진행

visited = [False] * N
num = N
ans = 0


def dfs(energy):
    global visited, N, W, num, ans

    if num == 2:
        ans = max(ans, energy)
        return

    for i in range(1, N - 1):
        if not visited[i]:
            num -= 1
            visited[i] = True
            lp = i - 1
            rp = i + 1
            while visited[lp]:
                lp -= 1
            while visited[rp]:
                rp += 1
            dfs(energy + W[lp] * W[rp])
            visited[i] = False
            num += 1


dfs(0)

print(ans)
